contracts/Registry.sol

```diff
@@ -99,12 +99,35 @@ contract Registry is AccessControlDefaultAdminRules {
     );
   }
 
-  function registerReferral(
-    string calldata referrerId,
-    string calldata protocolId
-  ) external {
-    _referrerIsRegistered = false;
-    for (uint256 i = 0; i < _referrerIdToProtocolIds[referrerId].length; i++) {
+  function getUserRegistrations(
+                                 address userAddress,
+                                 bytes32[] calldata protocolIds
+  ) external view returns (bool[] memory) {
+      bool[] memory registrations = new bool[](protocolIds.length);
+      for (uint256 i = 0; i < protocolIds.length; i++) {
+          registrations[i] = this.getUserRegistration(userAddress, protocolIds[i]);
+      }
+      return registrations;
+  }
+
+  function getUserRegistration(
+                                 address userAddress,
+                                 bytes32 protocolId
+  ) external view returns (bool) {
+      return _userInfoByProtocol[protocolId][userAddress].timestamp != 0;
+  }
+
+  // This registration functionality needs to be inside of an internal helper,
+  // rather than in the external `registerReferral` directly, since the batch
+  // `registerReferrals` must use this logic as a subroutine, and nested internal
+  // contract calls modify the value of `msg.sender`.
+  function _registerReferral(
+                             bytes32 referrerId,
+                             bytes32 protocolId,
+                             address userAddress
+  ) internal {
+      _referrerIsRegistered = false;
+      for (uint256 i = 0; i < _referrerIdToProtocolIds[referrerId].length; i++) {
       if (
         keccak256(abi.encodePacked(_referrerIdToProtocolIds[referrerId][i])) ==
         keccak256(abi.encodePacked(protocolId))
```

Comment:
We can now do this directly which is more efficient:
```suggestion
        _referrerIdToProtocolIds[referrerId][i] ==
        protocolId
```

Reply:
Oh :facepalm: missed this! Thanks for the catch